facebookmessenger-impl/src/main/java/ai/labs/facebookmessenger/endpoint/FacebookEndpoint.java

```diff
@@ -0,0 +1,265 @@
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.ws.rs.InternalServerErrorException;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.Callable;
+
+import static ai.labs.memory.model.SimpleConversationMemorySnapshot.SimpleConversationStep;
+import static ai.labs.memory.model.SimpleConversationMemorySnapshot.SimpleData;
+import static ai.labs.rest.restinterfaces.RestInterfaceFactory.RestInterfaceFactoryException;
+
+@RequestScoped
+@Slf4j
+public class FacebookEndpoint implements IFacebookEndpoint {
+    private static final String RESOURCE_URI_CHANNELCONNECTOR = "eddi://ai.labs.channel.facebook";
+    private final IBotStore botStore;
+    private final IHttpClient httpClient;
+    private final IJsonSerialization jsonSerialization;
+    private final IRestInterfaceFactory restInterfaceFactory;
+    private final String apiServerURI;
+    private final ICache<String, BotConfiguration> botConfigCache;
+    private final ICache<String, String> conversationIdCache;
+    private final ICache<String, MessengerClient> messengerClientCache;
+
+    @Inject
+    public FacebookEndpoint(IBotStore botStore,
+                            IHttpClient httpClient,
+                            IJsonSerialization jsonSerialization,
+                            IRestInterfaceFactory restInterfaceFactory,
+                            @Named("system.apiServerURI") String apiServerURI,
+                            ICacheFactory cacheFactory) {
+        this.botStore = botStore;
+        this.httpClient = httpClient;
+        this.jsonSerialization = jsonSerialization;
+        this.restInterfaceFactory = restInterfaceFactory;
+        this.apiServerURI = apiServerURI;
+        this.botConfigCache = cacheFactory.getCache("facebook.botConfiguration");
+        this.conversationIdCache = cacheFactory.getCache("facebook.conversationIds");
+        this.messengerClientCache = cacheFactory.getCache("facebook.messengerReceiveClients");
+    }
+
+    private MessengerClient getMessageClient(String botId) {
+        if (!messengerClientCache.containsKey(botId)) {
+            messengerClientCache.put(botId, createMessageClient(botId));
+        }
+
+        return messengerClientCache.get(botId);
+    }
+
+    private MessengerClient createMessageClient(String botId) {
+        BotConfiguration botConfiguration;
+        try {
+            Integer botVersion = botStore.getCurrentResourceId(botId).getVersion();
+            String cacheKey = botId + ":" + botVersion;
+            if (!botConfigCache.containsKey(cacheKey)) {
+                botConfiguration = botStore.read(botId, botVersion);
+                botConfigCache.put(cacheKey, botConfiguration);
+            } else {
+                botConfiguration = botConfigCache.get(cacheKey);
+            }
```

Comment:
Better use the following to avoid double lookup:
```java
if ((botConfiguration = botConfigCache.get(cacheKey)) == null) {
    botConfiguration = botStore.read(botId, botVersion);
    botConfigCache.put(cacheKey, botConfiguration);
}
```